### DISCUSSION TITLE
Fastify over Express

### DISCUSSION THREAD

**Discussion by juanpicado** [verdaccio/verdaccio#2155] [Ideas]
Since the fork from Sinopia verdaccio has been using `express` as default web application framework. That was almost 5 years ago. The project has been evolving slowly to a better structure and improve the maintenance. The web structure has barely changed due I was waiting some update from express project, which is great, but unfortunately is blocking me to improve Verdaccio in that area.

Since a while I've been thinking whether would worth move on and use something else, but it requires a lot of effort since the express middle-ware is quite embedded in the architecture all over the place (not to mention some plugins like `verdaccio-audit`). 

Also there are other concerns and I've haven't been able to address, either lack of time, by knowledge or help. 

### Callback and promises 

Verdaccio is callback based and some new implementations are already Promised based, I won't mention the benefits of promises, you can read [here](https://stackoverflow.com/questions/45041462/node-js-when-to-use-promises-vs-callbacks), but my preferred one is readability.

### Middleware

Other issue we have is currently the middleware plugins are being loaded withing the API namespace and bring issues like https://github.com/verdaccio/verdaccio/pull/1841 https://github.com/verdaccio/verdaccio/issues/1866 https://github.com/verdaccio/verdaccio/issues/1225 . I think we need to encapsulate the request and every middleware plugin would be able to apply any extra sauce they think is need it so other endpoints are not affected.

### Error handling

The current approach works like this, every time an endpoint thrown an error, the error is passed down like this

```
return next(ErrorCode.getCode(HTTP_STATUS.LOOP_DETECTED, 'loop detected'));
```

This is a normal behavior and documented [here](https://expressjs.com/en/guide/error-handling.html) which is fine, but  makes the current and [custom error handling](https://github.com/verdaccio/verdaccio/blob/master/src/api/middleware.ts#L181) a bit of a hell since the web and the api have different ways to reply.

### Other issues

I don't remember all details, but I'm sure there is more, maybe I'll add more in this section as long I found notes or piece of code that bring more details.

 ## My proposal

Is Fastify,  I'm not such an expert since have only few weeks playing around but seems like is the perfect candidate to make the job. I'm not a fan boy of libraries, I just want the best that works for the project in long term, is constantly updated, belongs to the OpenJS foundation, plugin based.

https://www.fastify.io/

### How to contribute

The project is evolving slowly to a new structure (https://github.com/verdaccio/verdaccio/tree/6.x) and `master` has the new Verdaccio 5 and is express based and should continue like this. But 6.x is flexible for breaking changes as long all test are green. 

* Contribute with a migration proposal
* A new middleware plugins structure

I'd like to discuss here any thoughts, if you want to discuss technical details, you can find me on http://chat.verdaccio.org `#contribute` channel.

# Work in Progress

- [x] Ping endpoint migrated https://github.com/verdaccio/verdaccio/commit/55ee3fdd97c52e6e58003eb5bf93c48764b9499c
-  [x] user (easy) https://github.com/verdaccio/verdaccio/pull/2301 WIP 
   -  [ ] migrate middleware auth (hard) - (token verify, legacy, JWT) WIP   
-  [x]  search (v1 only) ignore `/-/all` (high)  https://github.com/verdaccio/verdaccio/pull/2432 (uses `undici` + `undici-fetch`) 
-  [x] whoami endpoint #2599
-  [x] download tarball https://github.com/verdaccio/verdaccio/pull/2600

**Comment by Eomm** [comment]
Hi, Fastify old-user here 👋🏽 

I'm following the discussion so far, I think this would be a great to support Verdaccio evolving
I will try to draft a branch so, if someone would like to join in and go for the big part of the implementation

**Reply by juanpicado** [reply]
I can join after the rollout v5 at ~middle of April, my plan is the branch 6.x becomes master and current master 5.x branch.

**Reply by hoghweed** [reply]
I too would like to join. Let me know how can I contribute and join in that sense.

**Comment by juanpicado** [comment]
The master branch contains an experimental server which runs like this
```
pnpm install
pnpm build
pnpm debug:fastify
```
or 

```
npm i -g verdaccio@6-next
verdaccio fastify-server
```

This will run a simple fastify server throug the CLI, from that point I still need to figure out next steps.

**Comment by hoghweed** [comment]
@juanpicado thanks for what you wrote above. Considering I'm not (yet) aware of Verdaccio internals is a wonderful starting point to know your point of view. 

What you wrote is part (not at all because it differs from standard porting due to Verdaccio peculiarities) of what usually is considered to move to fastify. 

I have some thoughts about a possible base roadmap, this means it would be enriched while get familiar with Verdaccio sources and internals. 

How can I start to write down such a migration proposal roadmap? I was considering also having a board within projects mapping a migration proposal & plan.

**Reply by juanpicado** [reply]
> I have some thoughts about a possible base roadmap, this means it would be enriched while get familiar with Verdaccio sources and internals.
> How can I start to write down such a migration proposal roadmap? I was considering also having a board within projects mapping a migration proposal & plan.

Nice, I did create a project

https://github.com/verdaccio/verdaccio/projects/17?add_cards_query=is%3Aopen

Not sure which permissions are need it for it, but I'll figure it out.

I can help describing here a bit the current status to provide more context.

**Reply by hoghweed** [reply]
Please, do it. Describing here the current status would help a lot. a HUGE help!

**Reply by juanpicado** [reply]
I'll do it along the week 👍🏼 , it is gonna be long writing.

**Comment by juanpicado** [comment]
Hi, I'm trying to describe here the current status.

We have 3 development branches:
- ~[4.x](https://github.com/verdaccio/verdaccio/tree/4.x): Legacy, just security patches~ <-- not longer legacy
- [5.x](https://github.com/verdaccio/verdaccio/tree/5.x): Latest, I'm not accepting features, just minor bugs or security patches.
- `master` which is intended to be `6.x` eventually and where I'd like to have `Fastify`.

For getting started I merged https://github.com/verdaccio/verdaccio/pull/2201 which introduce a new package which start a basic fastify server. For running just.

```
pnpm install
pnpm build
pnpm debug -- new
```

The project on master is based on modules where every package should have a clear scope, everything starts on `@verdaccio/cli` which is based on `clipanion` and ad new commands and is where the new server start `newServet.s`. 

The project uses the latest release of express and from that point follows it philosophy, middle-wares for the api `"@verdaccio/api"` layer (with all api routes) and another layer for web `"@verdaccio/web"` and both are combined on `"@verdaccio/server"` where is the module should be replaced by the new server fastify `"@verdaccio/fastify-migration"` <- I just added a meaningful name in the meantime.

Diving more into the modules, the server creates instances of `'@verdaccio/auth'` and `'@verdaccio/store'` which are self-explanatory and should not have any reference from fastify or express. In theory those packages should not change.

I hope this helps to understand how the server side works.

**Reply by amitgilad3** [reply]
Hi @juanpicado ,since i have been working with fastify for some time i would like to give it a try and move verdaccio to fastify. 

from what i understand , in order to move i will use the new package @verdaccio/fastify-migration and start migrating the:

- [ ] @verdaccio/server 
- [ ]  @verdaccio/api 
- [ ]  @verdaccio/web
- [ ]  @verdaccio/auth
- [ ]  @verdaccio/store

wdyt ??

**Reply by Eomm** [reply]
I was looking on it too and what I understood is:

```
// entry point: build the server and it should not change
packages\node-api\src\server.ts

// the express.js server builder
packages\server\src\server.ts (defineAPI)
  --> Manage CORS, rate-limit, 4xx and 5xx errors, payload compressions, debuging endpoints
  --> Storage plugins should not change
  --> Auth middleware may change
  --> Plugin middlewares may change
  
  // register the npm API endpoints
  --> packages\api\src\index.ts
  // register the web UI
  --> packages\web\src\web-middleware.ts

// the fastify migration
packages\core\server\src\server.ts
```

I was trying the [`fastify-express`](https://github.com/fastify/fastify-express) plugin that should works if there is not too much monkey patching

**Reply by juanpicado** [reply]
Yes @Eomm that's more or less the current status.

@amitgilad3  I've added the new module in order do not break the current functionality, in other words, the new Fastify server should run consuming other modules, the current server, express since is pretty stable the migration should be smooth as possible and both servers should continuously working. Some packages are quite "experssified" or middleware based (maybe `fastify-express` helps there), I am not aware of pro and cons yet, but if that speed up the migration, that's very welcome.

I think some packages might need to duplicate code (classes, etc) or maybe create new modules and that's fine. I'd prefer that rather refactor code and compromise releases since this migration might take a while or not, who knows and it's fine to ship a parallel server even if is incomplete, nobody will use it until is ready anyway, and for the other side keep test green is really important to me.

There are some easy endpoints could be an easy start, we have two big groups.

- API (cli)
  - user (easy) https://github.com/verdaccio/verdaccio/pull/2301
  - packages (high)
      - metadata
      - ~tarball~ https://github.com/verdaccio/verdaccio/pull/2600       
  - dist-tags ( medium)
  - ~ping (easy)~
  - ~whoami (easy)~ #2599
  - stars (medium)
  - search (v1 only) ignore `/-/all` (high)  https://github.com/verdaccio/verdaccio/pull/2432 (uses `undici` + `undici-fetch`)
  - publish  (high)
-  Web 
   - readme  (easy) #2613
   - packages  ( medium) 
   - user (easy) #2624
   - sidebar  (easy) #2618 

I'd start without authentication layer to make it easier, app endpoints are designed to work without tokens, could be added later on, Am I wrong here?, not 100% sure. 

I'd focus on migrate basic functionality (easy endpoints) and progressively either migrating test or creating new in case there aren't. Most of the test uses `supertest` and the new ones also uses `nock` which is quite powerful and easy to migrate if move to fastify and is basically integration tests and unit test only for small methods or small-medium size.

I haven't had the time to dive too much in decorators, hooks, so my ideas are still high level. 

Something else to share, if you have noticed the app is `callback` based due legacy, I've been moving to `async/await` in the places that make sense (this task is also at the main roadmap), so create parallel code might be easier rather migrate some spaghetti is around.

Those are my main thoughts, hopefully sound reasonable.

**Comment by juanpicado** [comment]
Free course how to migrate from express to Fastify  18th May 

https://twitter.com/simonplend/status/1391760477433257984?s=21

**Comment by juanpicado** [comment]
@dianmorales will help on this topic.

https://github.com/verdaccio/verdaccio/issues/2623
